### PR TITLE
[MTM-60271] OAI-Secure tokens generated in BODY will be limited based…

### PR DESCRIPTION
**Pull request by mbak-c8y**
The user can generate a JWT OAI-Secure token in two ways:
-In cookies, which are commonly used by UI applications.
-In the JSON body, which is used mostly by services, which does not use cookies.

One of the customers generates tokens per request, which floods user representation.
Tokens generated in cookie format are limited by the sessionConfiguration:maximumNumberOfParallerSession.
This change is about limiting tokens generated in body format in the same manner as it is done for cookie tokens.

So if the next generated session (with token) exceeds a maximum number of parallel sessions then the oldest session with matching tokens is terminated.
Fix versions: 
10.18.540.186
10.18.0.467
10.20.525.0

### CODE REVIEW

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+component:
+  - value: q3kclF6pO
+    label: Authentication
+build_artifact:
+  - value: tc-QHwMfWtBk7
+    label: cumulocity
+ticket: MTM-60271
+version: 10.20.525.0
+---
+The maximum number of OAI-Secure Tokens generated by the "/tenant/oauth/token" API will be limited according to the `Maximum parallel sessions per user` in session configuration.
```
Comment (BeateRixen) [inline content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:17]:
```suggestion
The maximum number of OAI-Secure tokens generated by the `/tenant/oauth/token` API is now limited according to the `Maximum parallel sessions per user` parameter in the session configuration.
```

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+---
+date: ""
+title: Limit number of tokens generated in body based on session configuration
```
Comment (BeateRixen) [inline content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:3]:
```suggestion
title: Limited number of tokens generated in body based on the session configuration
```

Comment (BeateRixen) [reply content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:3]: Here it says "in body", below it says "in API". Phrasing should be consistent. Also, should it be mentioned that tokens generated by cookies have already be limited?

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+component:
+  - value: q3kclF6pO
+    label: Authentication
+build_artifact:
+  - value: tc-QHwMfWtBk7
+    label: cumulocity
+ticket: MTM-60271
+version: 10.20.525.0
+---
+The maximum number of OAI-Secure tokens generated in body by the `/tenant/oauth/token` API is now limited according to the `Maximum parallel sessions per user` parameter in the session configuration.
```
Comment (BeateRixen) [inline content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:17]:
```suggestion
The maximum number of OAI-Secure tokens generated in the body by the `/tenant/oauth/token` API is now limited according to the `Maximum parallel sessions per user` parameter in the session configuration.
```

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md (hunk 2 above)
Comment (mbak-c8y) [reply content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md]: Imo I would rather skip this information. In the case of tokens generated in the cookies, tokens are not limited but rather sessions, so to not complicate this release notes, I will skip this information. This is already in the documentation described.

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
+---
+date: ""
+title: Limited number of tokens generated in body based on the session configuration
```
Comment (BeateRixen) [inline content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:3]:
```suggestion
title: Limited number of tokens generated in the body based on the session configuration
```

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
+---
+date: ""
+title: Limited number of tokens generated in the body based on the session configuration
```
Comment (BeateRixen) [inline content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:3]:
```suggestion
title: Number of tokens generated in the body is now limited in the session configuration
```

**Review by BeateRixen — APPROVED** [review]

**Review by pink-mongoose — COMMENTED** [review]
Is there a security enhancement aspect to this?  It feels like we should be able to say that we have improved security by limiting the number of tokens that can be generated.

**Comment by pink-mongoose** [comment]
> > Is there a security enhancement aspect to this? It feels like we should be able to say that we have improved security by limiting the number of tokens that can be generated.
> 
> Probably we can say that this is a security enhancement, but rather, protecting clients from flooding the database with unused tokens.

So, would the following make sense:
To enhance OAI-Secure token management, the number of tokens generated in the JSON body will be limited to the maximum configured in the sessionConfiguration

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+component:
+  - value: q3kclF6pO
+    label: Authentication
+build_artifact:
+  - value: tc-QHwMfWtBk7
+    label: cumulocity
+ticket: MTM-60271
+version: 10.20.525.0
+---
+To enhance OAI-Secure token management, the number of tokens generated in the JSON body will be limited to the maximum configured in the sessionConfiguration.
```
Comment (BeateRixen) [inline content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:17]:
```suggestion
To enhance the OAI-Secure token management, the number of tokens generated in the JSON body will be limited to the maximum configured in the session configuration.
```

File: content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+---
+date: ""
+title: Number of tokens generated in the body is now limited in the session configuration
```
Comment (BeateRixen) [inline content/change-logs/platform-services/cumulocity-10-20-525-0-limit-number-of-tokens-generated-in-body-based-on-session-configuration.md:3]:
```suggestion
title: Number of tokens generated in the JSON body is now limited
```

**Review by BeateRixen — APPROVED** [review]

**Comment by BeateRixen** [comment]
@mbak-c8y Can be merged from my side. Please create a second PR for 2024 (with the version 10.18.540.186) and add the text to the Jira ticket for 10.18 (done by me).

**Comment by mbak-c8y** [comment]
> @mbak-c8y Can be merged from my side. Please create a second PR for 2024 (with the version 10.18.540.186) and add the text to the Jira ticket for 10.18 (done by me).

Done: https://github.com/SoftwareAG/c8y-docs/pull/2366

**Pull request merged**